pkg/gatewayconfigurationserver/middleware.go

```diff
@@ -15,39 +15,60 @@
 package gatewayconfigurationserver
 
 import (
-	echo "github.com/labstack/echo/v4"
+	"context"
+	"net/http"
+
+	"github.com/gorilla/mux"
 	"go.thethings.network/lorawan-stack/v3/pkg/auth/rights"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
 	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/webhandlers"
+	"go.thethings.network/lorawan-stack/v3/pkg/webmiddleware"
+	"google.golang.org/grpc/metadata"
 )
 
-const gatewayIDKey = "gateway_id"
+func (s *Server) fillRequestContext(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := s.FillContext(r.Context())
+		ctx = log.NewContextWithField(ctx, "namespace", "gatewayconfigurationserver")
 
-func (gcs *GatewayConfigurationServer) validateAndFillIDs() echo.MiddlewareFunc {
-	return func(next echo.HandlerFunc) echo.HandlerFunc {
-		return func(c echo.Context) error {
-			ctx := gcs.getContext(c)
-			gtwID := ttnpb.GatewayIdentifiers{
-				GatewayID: c.Param(gatewayIDKey),
-			}
-			if err := gtwID.ValidateContext(ctx); err != nil {
-				return err
-			}
-			c.Set(gatewayIDKey, gtwID)
+		md := metadata.New(map[string]string{
+			"authorization": r.Header.Get("Authorization"),
+		})
+		if ctxMd, ok := metadata.FromIncomingContext(ctx); ok {
+			md = metadata.Join(ctxMd, md)
+		}
+
+		next.ServeHTTP(w, r.WithContext(metadata.NewIncomingContext(ctx, md)))
```

Comment:
Nit: Better to write `ctx = metadata.NewIncomingContext(ctx, md)` above and only write `r.WithContext(ctx)` here. That would make it easier to add other "fillers".